docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+    IL_000a: stloc.0
+    IL_000b: ldloca.s 0
+    IL_000d: call instance bool [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter::get_IsCompleted()
+    IL_0012: brtrue.s IL_001b
+
+    IL_0014: ldloc.0
+    IL_0015: call class [System.Runtime]System.Threading.Tasks.Task System.Runtime.CompilerServices.RuntimeHelpers::AwaitAwaiterFromRuntimeAsync<valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter>(!!0)
+    IL_001a: pop
+
+    IL_001b: ldloca.s 0
+    IL_001d: call instance void [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter::GetResult()
+    IL_0022: ret
+}
+```
+
+#### Await local of concrete type `Task<T>`
+
+```cs
+class C
+{
+    static Task<int> M();
+}
+
+var local = M();
+var i = await local;
+```
+
+Translated C#:
+
+```cs
+var local = C.M();
+var i =
+{
+    var awaiter = local.GetAwaiter();
+    if (!awaiter.IsComplete)
+    {
+        /* Runtime-Async Call */ System.Runtime.CompilerServices.RuntimeHelpers.AwaitAwaiterFromRuntimeAsync<System.Runtime.CompilerServices.TaskAwaiter>(awaiter);
+    }
+    awaiter.GetResult()
+};
+```
+
+```il
+{
+    .locals init (
+        [0] valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32> awaiter
+    )
+
+    IL_0000: call class [System.Runtime]System.Threading.Tasks.Task`1<int32> C::M()
+    IL_0005: callvirt instance valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<!0> class [System.Runtime]System.Threading.Tasks.Task`1<int32>::GetAwaiter()
+    IL_000a: stloc.0
+    IL_000b: ldloca.s 0
+    IL_000d: call instance bool valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32>::get_IsCompleted()
+    IL_0012: brtrue.s IL_001b
+
+    IL_0014: ldloc.0
+    IL_0015: call class [System.Runtime]System.Threading.Tasks.Task System.Runtime.CompilerServices.RuntimeHelpers::AwaitAwaiterFromRuntimeAsync<valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32>>(!!0)
+    IL_001a: pop
+
+    IL_001b: ldloca.s 0
+    IL_001d: call instance !0 valuetype [System.Runtime]System.Runtime.CompilerServices.TaskAwaiter`1<int32>::GetResult()
+    IL_0022: pop
+    IL_0023: ret
+}
+```
+
+#### Await a `T`-returning method
+
+```cs
+class C
+{
+    static T M<T>();
+}
+
+await C.M<Task>();
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a generic `T` `Task<T>`-returning method
+
+```cs
+class C
+{
+    static Task<T> M<T>();
+}
+
+await C.M<int>();
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a `Task`-returning delegate
+
+```cs
+delegate Task AsyncDelegate();
+
+class C
+{
+    static Task M();
+}
+
+AsyncDelegate d = C.M;
+await d;
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a `T`-returning delegate
+
+```cs
+
+class C
+{
+    static Task M();
+}
+
+Func<Task> d = C.M;
+await d;
```

Comment:
I didn't follow. Why is this expected to work?
This would not work in compiler-async today: [sharplab](https://sharplab.io/#v2:D4AQDABCCMCsDcBYAUCATNFKSwDwgDYA+CAEwgF4IBhAOgFklUBOMp7NGlAbxQn6gBmKNAJQx9ABQBKSiQAuACwBOAewDuEAHYBXADZ6mAXyzIgA)